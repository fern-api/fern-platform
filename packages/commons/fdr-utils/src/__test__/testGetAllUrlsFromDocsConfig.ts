import { DocsV2Read } from "@fern-api/fdr-sdk";
import fs from "fs";
import path from "path";
import { getAllUrlsFromDocsConfig } from "../getAllUrlsFromDocsConfig";

export function testGetAllUrlsFromDocsConfig(fixtureName: string): void {
    // eslint-disable-next-line vitest/valid-title
    describe(fixtureName, () => {
        it("gets all urls from docs config", async () => {
            const fixturePath = path.join(__dirname, "fixtures", `${fixtureName}.json`);

            const content = fs.readFileSync(fixturePath, "utf-8");

            const fixture = JSON.parse(content) as DocsV2Read.LoadDocsForUrlResponse;

            const urls = getAllUrlsFromDocsConfig(
                fixture.baseUrl.domain,
                fixture.baseUrl.basePath,
                fixture.definition.config.navigation,
                fixture.definition.apis,
            );

            expect(urls).toMatchSnapshot();
        });
    });
}
