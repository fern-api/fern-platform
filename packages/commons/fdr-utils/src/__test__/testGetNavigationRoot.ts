import { DocsV2Read } from "@fern-api/fdr-sdk";
import fs from "fs";
import path from "path";
import { getNavigationRoot } from "../getNavigationRoot";

export function testGetNavigationRoot(fixtureName: string, slug: string): void {
    // eslint-disable-next-line vitest/valid-title
    describe(fixtureName, () => {
        it("gets navigation root for /" + slug, async () => {
            const fixturePath = path.join(__dirname, "fixtures", `${fixtureName}.json`);

            const content = fs.readFileSync(fixturePath, "utf-8");

            const fixture = JSON.parse(content) as DocsV2Read.LoadDocsForUrlResponse;

            const urls = getNavigationRoot(
                slug.split("/"),
                fixture.baseUrl.basePath,
                fixture.definition.apis,
                fixture.definition.config.navigation,
            );

            expect(stripForSnapshot(urls)).toMatchSnapshot();
        });
    });
}

function stripForSnapshot(obj: ReturnType<typeof getNavigationRoot>): unknown {
    if (obj == null || obj.type === "redirect") {
        return obj;
    }

    // only test the tab index, version index, and current node
    return {
        type: obj.type,
        found: {
            currentTabIndex: obj.found.currentTabIndex,
            currentVersionIndex: obj.found.currentVersionIndex,
            currentNode: obj.found.currentNode,
        },
    };
}
