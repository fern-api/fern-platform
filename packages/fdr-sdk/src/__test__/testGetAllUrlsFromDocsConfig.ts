import urljoin from "url-join";
import { FernNavigation } from "..";
import { NodeCollector } from "../navigation/NodeCollector";

export function testGetAllUrlsFromDocsConfig(
  root: FernNavigation.RootNode,
  domain: string
): void {
  it("gets all urls from docs config", async () => {
    const collector = NodeCollector.collect(root);
    const urls = collector.staticPageSlugs.map((slug) => urljoin(domain, slug));
    expect(urls).toMatchSnapshot();
  });
}
