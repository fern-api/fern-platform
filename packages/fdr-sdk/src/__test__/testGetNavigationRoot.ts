import { FernNavigation } from "..";
import { findNode } from "../navigation/utils";

export function testGetNavigationRoot(root: FernNavigation.RootNode, slug: string): void {
    it("gets navigation root for /" + slug, async () => {
        const found = findNode(root, FernNavigation.slugjoin(slug));

        if (found.type === "found") {
            expect(found.node).toMatchSnapshot();
            expect(found.currentVersion?.versionId).toMatchSnapshot();
            expect(found.currentTab?.slug).toMatchSnapshot();
        } else {
            expect(found).toMatchSnapshot();
        }
    });
}
