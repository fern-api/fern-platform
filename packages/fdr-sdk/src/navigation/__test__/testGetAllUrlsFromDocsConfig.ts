import urljoin from "url-join";
import { FernNavigation } from "../..";
import { NodeCollector } from "../NodeCollector";

export function testGetAllUrlsFromDocsConfig(root: FernNavigation.RootNode, domain: string): void {
    it("gets all urls from docs config", async () => {
        const slugCollector = NodeCollector.collect(root);
        const urls = slugCollector.getPageSlugs().map((slug) => urljoin(domain, slug));
        expect(urls).toMatchSnapshot();
    });
}
