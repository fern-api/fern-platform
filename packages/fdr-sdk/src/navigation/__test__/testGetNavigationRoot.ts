import fs from "fs";
import path from "path";
import { DocsV2Read } from "../../client";
import { convertLoadDocsForUrlResponse, findNode, slugjoin } from "../utils";

export function testGetNavigationRoot(fixtureName: string, slug: string): void {
    // eslint-disable-next-line vitest/valid-title
    describe(fixtureName, () => {
        it("gets navigation root for /" + slug, async () => {
            const fixturePath = path.join(__dirname, "fixtures", `${fixtureName}.json`);

            const content = fs.readFileSync(fixturePath, "utf-8");

            const fixture = JSON.parse(content) as DocsV2Read.LoadDocsForUrlResponse;
            const node = convertLoadDocsForUrlResponse(fixture);
            const found = findNode(node, slugjoin(slug));

            if (found.type === "found") {
                expect(found.node).toMatchSnapshot();
                expect(found.currentVersion?.versionId).toMatchSnapshot();
                expect(found.currentTab?.slug).toMatchSnapshot();
            } else {
                expect(found).toMatchSnapshot();
            }
        });
    });
}
