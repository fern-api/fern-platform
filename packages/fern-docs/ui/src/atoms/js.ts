import type { DocsV1Read } from "@fern-api/fdr-sdk";
import type { JsConfig } from "@fern-docs/ui";
import { isEqual } from "es-toolkit/predicate";
import { selectAtom } from "jotai/utils";
import { DOCS_ATOM } from "./docs";

export const JS_ATOM = selectAtom(DOCS_ATOM, (docs) => docs.js, isEqual);

export function withCustomJavascript(
  readShapeJsConfig: DocsV1Read.JsConfig | undefined,
  resolveFileSrc: (fileId: string) => { src: string } | undefined
): JsConfig | undefined {
  const remote = [
    ...(readShapeJsConfig?.remote ?? []),
    ...(readShapeJsConfig?.files ?? []).map((file) => ({
      url: resolveFileSrc(file.fileId)?.src,
      strategy: file.strategy,
    })),
  ].filter(isRemote);

  const toRet = {
    inline: readShapeJsConfig?.inline,
    remote: remote.length > 0 ? remote : undefined,
  };

  if (!toRet.inline && !toRet.remote) {
    return undefined;
  }

  return toRet;
}

type RemoteJs = NonNullable<JsConfig["remote"]>[number];

function isRemote(remote: {
  url: string | undefined; // potentially undefined if the file is not found
  strategy: RemoteJs["strategy"];
}): remote is RemoteJs {
  return remote.url != null;
}
