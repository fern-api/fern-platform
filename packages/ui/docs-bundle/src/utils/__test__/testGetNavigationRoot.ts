import { DocsV2Read } from "@fern-api/fdr-sdk";
import fs from "fs";
import path from "path";
import { getNavigationRoot } from "../getNavigationRoot";

// eslint-disable-next-line jest/no-export
export function testGetNavigationRoot(fixtureName: string, slug: string): void {
    // eslint-disable-next-line jest/valid-title
    describe(fixtureName, () => {
        it("gets navigation root for " + slug, async () => {
            const fixturePath = path.join(__dirname, "fixtures", `${fixtureName}.json`);

            const content = fs.readFileSync(fixturePath, "utf-8");

            const fixture = JSON.parse(content) as DocsV2Read.LoadDocsForUrlResponse;

            const urls = getNavigationRoot(
                slug.split("/"),
                fixture.baseUrl.basePath,
                fixture.definition.apis,
                fixture.definition.config.navigation,
            );

            expect(urls).toMatchSnapshot();
        });
    });
}
