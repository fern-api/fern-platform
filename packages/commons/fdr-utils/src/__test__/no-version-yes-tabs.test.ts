import { testGetAllUrlsFromDocsConfig } from "./testGetAllUrlsFromDocsConfig.js";
import { testGetNavigationRoot } from "./testGetNavigationRoot.js";

const FIXTURE = "no-version-yes-tabs";

describe("getAllUrlsFromDocsConfig", () => {
    testGetAllUrlsFromDocsConfig(FIXTURE);
});

describe("getNavigationRoot", () => {
    testGetNavigationRoot(FIXTURE, "docs/api");
    testGetNavigationRoot(FIXTURE, "docs/api/tab-1");
    testGetNavigationRoot(FIXTURE, "docs/api/section-1");
    testGetNavigationRoot(FIXTURE, "docs/api/page-2");
    testGetNavigationRoot(FIXTURE, "docs/api/tab-2");
    testGetNavigationRoot(FIXTURE, "docs/api/tab-2/section-5");
});
