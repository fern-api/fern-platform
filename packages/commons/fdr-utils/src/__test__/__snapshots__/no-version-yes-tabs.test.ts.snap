// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllUrlsFromDocsConfig no-version-yes-tabs gets all urls from docs config 1`] = `
[
  "/docs/api",
  "/docs/api/tab-1",
  "/docs/api/tab-1/section-1",
  "/docs/api/tab-1/section-1/page-1",
  "/docs/api/page-2",
  "/docs/api/tab-1/section-2",
  "/docs/api/section-2",
  "/docs/api/tab-1/section-2/page-4",
  "/docs/api/tab-1/page-5",
  "/docs/api/tab-2",
  "/docs/api/tab-2/page-7",
  "/docs/api/tab-2/page-8",
  "/docs/api/tab-2/section-5",
]
`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api 1`] = `
{
  "redirect": "/docs/api/tab-1/section-1/page-1",
  "type": "redirect",
}
`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api/page-2 1`] = `
{
  "found": {
    "currentNode": {
      "description": undefined,
      "fullSlug": [
        "page-2",
      ],
      "id": "page-2.md",
      "slug": [
        "docs",
        "api",
        "page-2",
      ],
      "title": "Page 2",
      "type": "page",
      "urlSlug": "page-2",
    },
    "currentTabIndex": 0,
    "currentVersionIndex": -1,
  },
  "type": "found",
}
`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api/section-1 1`] = `undefined`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api/tab-1 1`] = `
{
  "found": {
    "currentNode": {
      "description": undefined,
      "fullSlug": [
        "tab-1",
      ],
      "id": "page-6.md",
      "slug": [
        "docs",
        "api",
        "tab-1",
      ],
      "title": "Page 6",
      "type": "page",
      "urlSlug": "page-6",
    },
    "currentTabIndex": 0,
    "currentVersionIndex": -1,
  },
  "type": "found",
}
`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api/tab-2 1`] = `
{
  "found": {
    "currentNode": {
      "description": undefined,
      "fullSlug": [
        "tab-2",
      ],
      "id": "page-10.md",
      "slug": [
        "docs",
        "api",
        "tab-2",
      ],
      "title": "Page 10",
      "type": "page",
      "urlSlug": "page-10",
    },
    "currentTabIndex": 1,
    "currentVersionIndex": -1,
  },
  "type": "found",
}
`;

exports[`getNavigationRoot no-version-yes-tabs gets navigation root for /docs/api/tab-2/section-5 1`] = `
{
  "redirect": "/docs/api/tab-1",
  "type": "redirect",
}
`;
